app/Http/Controllers/RedirectController.php

```diff
@@ -4,21 +4,71 @@
 
 namespace App\Http\Controllers;
 
+use App\ModelFunctions\SessionFunctions;
+use App\Models\Album;
+use Illuminate\Http\Request;
+use Illuminate\Support\Facades\Hash;
+
 class RedirectController extends Controller
 {
-	/**
-	 * Trivial redirection.
-	 */
-	public function album($albumid)
-	{
-		return redirect('gallery#' . $albumid);
-	}
-
-	/**
-	 * Trivial redirection.
-	 */
-	public function photo($albumid, $photoid)
-	{
-		return redirect('gallery#' . $albumid . '/' . $photoid);
-	}
+
+        /**
+         * @var SessionFunctions
+         */
+        private $sessionFunctions;
+
+        /**
+         * @param SessionFunctions    $sessionFunctions
+         */
+        public function __construct(
+                SessionFunctions $sessionFunctions
+        ) {
+                $this->sessionFunctions = $sessionFunctions;
+        }
+
+        /**
+         * Trivial redirection.
+         *
+         * @param Request $request
+         * @param string $albumid
+         *
+         */
+        public function album(Request $request, $albumid)
+        {
+                if (!$this->sessionFunctions->has_visible_album($albumid) && $request['password'] != '') {
+                        $this->unlockPasswordAlbum($request['password']);
+                }
+                return redirect('gallery#' . $albumid);
+        }
+
+        /**
+         * Trivial redirection.
+         *
+         * @param Request $request
+         * @param string $albumid
+         * @param string $photoid
+         */
+        public function photo(Request $request, $albumid, $photoid)
+        {
+                if (!$this->sessionFunctions->has_visible_album($albumid) && $request['password'] != '') {
+                        $this->unlockPasswordAlbum($request['password']);
+                }
+                return redirect('gallery#' . $albumid . '/' . $photoid);
+        }
+
+        /**
+         * Provided an password, add all the albums that the password unlocks
+         *
+         * @param string $password
+         */
+        private function unlockPasswordAlbum(string $password) {
+                $albums = Album::whereNotNull('password')->where('password', '!=', '')->get();
```

Comment:
No, this is way too dangerous.
You know the album ID, thus you should query the album, check if the password is Ok,

And if it is a the correct password, then we also do the check for the other albums.

It is a bit slower but this increases the difficulty of unlocking all the albums by bruteforce.

Reply:
Given a photo uploaded to a password protected album.
I want this photo to be available for people opening the smart album `recent` with the following link: `https://domain.com/r/recent?password=my_password`.

Reply:
So I assume that your `recent` smart album is public ?

Reply:
Yes

Reply:
Just to clarify, you want to add a password to your `recent` folder or you want your `recent` folder to display recent pictures in the "_unlocked_" folders?

Reply:
I want the `recent` smart album to display recent pictures from the "unlocked" folders (and public photos).
And, I only want provide one share link to the `recent` smart album.

Reply:
> Just to clarify, you want to add a password to your recent folder or you want your recent folder to display recent pictures from the "unlocked" folders?

`recent` smart album currently works this way, if you unlock a password protected album in a given session. Then when this session browse the `recent` smart album, he will see the pictures from the "unlocked" folders.

Reply:
Hooo I forgot about that.